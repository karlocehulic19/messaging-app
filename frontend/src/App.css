form {
  display: grid;
  background-color: var(--tertiary-color);
  padding: 1em;
  color: var(--secondary-color);
  border-radius: 1em;
  gap: 1em;
}

a {
  color: rgb(52, 142, 216);
  font-weight: 600;
  cursor: pointer;
}

a:hover {
  text-decoration: underline;
}

.dropbox {
  text-align: center;
  border-radius: 0.5em;
  border: 0.2em;
  border-style: dashed;
}

input {
  background-color: rgba(49, 54, 63, 0.61);
  color: var(--quaternary-color);
  padding: 0.6em 0.6em;
  border-radius: 0.6em;
  border: 2px solid rgb(0, 0, 0);
}

input::placeholder {
  color: rgb(236, 236, 236);
}

.error-popup {
  position: fixed;
  top: 5vh;
  background-color: rgba(254, 0, 0, 0.363);
  color: white;
  font-size: 1.2rem;
  padding: 0.5rem;
  border-radius: 0.5rem;
  animation: slide-down 0.05s ease-out;
  left: 50%;
  transform: translate(-50%, 0);
}

@keyframes slide-down {
  0% {
    top: 0px;
  }

  100% {
    top: 5vh;
  }
}

button {
  border: none;
  background-color: rgb(236, 236, 236);
  border-radius: 0.5em;
  padding: 0.3em;
  transition: all 0.1s;
}

button:enabled:hover {
  background-color: var(--secondary-color);
  color: var(--quaternary-color);
}

.profile-picture {
  border-radius: 100%;
}

button:hover {
  cursor: pointer;
}
