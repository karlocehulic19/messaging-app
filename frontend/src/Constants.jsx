const prod = {
  url: {
    BACKEND_URL: import.meta.env.VITE_PROD_BACKEND_URL,
  },
};

const dev = {
  url: {
    BACKEND_URL: import.meta.env.VITE_DEV_BACKEND_URL,
  },
};

export const config = import.meta.env.MODE === "production" ? prod : dev;
