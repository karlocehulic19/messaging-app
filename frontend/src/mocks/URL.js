global.URL.createObjectURL = vi.fn(() => "mockedFilePath");
global.URL.revokeObjectURL = vi.fn(() => null);
