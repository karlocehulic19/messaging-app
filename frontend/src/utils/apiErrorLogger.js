export default async function apiErrorLogger(error) {
  let resJSON;
  if (
    error.response &&
    error.response.headers.get("Content-Type").includes("application/json")
  )
    resJSON = await error.response.json();
  if (resJSON) console.error(resJSON);
  console.error(error);

  return resJSON;
}
