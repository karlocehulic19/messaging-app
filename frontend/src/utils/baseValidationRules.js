import { isAscii } from "validator";
import { isRequired } from "./validationExpressions";
import isEmail from "validator/lib/isEmail";

export const usernameRules = (ValidationBuilder, field = "username") => {
  ValidationBuilder.field(field)
    .addRule(isRequired, "Username field is required")
    .addRule(isAscii, "Invalid characters are provided");
};

export const emailRules = (ValidationBuilder, field = "email") => {
  ValidationBuilder.field(field)
    .addRule(isRequired, "Email field is required")
    .addRule(isEmail, "Value provided must be valid email address");
};
