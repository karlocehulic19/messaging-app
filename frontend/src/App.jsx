import "./App.css";
import { Routes, Route, BrowserRouter } from "react-router-dom";
import Login from "./pages/Login";
import Register from "./pages/Register";
import AuthProvider from "./contexts/AuthProvider";
import NonAuthRoute from "./routes/NonAuthRoute";
import PrivateRoute from "./routes/PrivateRoute";
import Main from "./pages/Main";
import Logout from "./pages/Logout";
import Settings from "./pages/Settings";
import PropTypes from "prop-types";

function App({
  routerRender = (children) => <BrowserRouter>{children}</BrowserRouter>,
}) {
  return (
    <>
      {routerRender(
        <AuthProvider>
          <Routes>
            <Route element={<NonAuthRoute />}>
              <Route path="/login" element={<Login />} />
              <Route path="/register" element={<Register />} />
            </Route>
            <Route element={<PrivateRoute />}>
              <Route path="/logout" element={<Logout />} />
              <Route path="/settings" element={<Settings />} />
              <Route path=":receiverUsername?" element={<Main />}></Route>
            </Route>
          </Routes>
        </AuthProvider>
      )}
    </>
  );
}

App.propTypes = {
  routerRender: PropTypes.func,
};

export default App;
