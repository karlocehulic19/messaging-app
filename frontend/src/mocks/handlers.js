import { http, HttpResponse } from "msw";
import { config } from "../Constants";

export const handlers = [
  http.post(`${config.url.BACKEND_URL}/login`, async ({ request }) => {
    const body = await request.json();
    if (!body.username || !body.password) {
      return new HttpResponse("Missing credentials", { status: 401 });
    }

    if (body.username != "someUsername" || body.password != "somePassword") {
      return HttpResponse.json(
        { messages: ["Username or password is incorrect"] },
        { status: 401 }
      );
    }
    return HttpResponse.json({ token: "randomJWTtoken" }, { status: 200 });
  }),
];
