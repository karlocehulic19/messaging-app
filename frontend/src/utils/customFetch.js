import { config } from "../Constants";
import { merge } from "lodash";

export class ResponseError extends Error {
  constructor(message, response) {
    super(message);
    this.response = response;
  }
}

function getDefaultOptions() {
  const defaultOptions = { headers: { "Content-Type": "application/json" } };
  const token = localStorage.getItem("site");

  if (token) {
    defaultOptions.headers.Authorization = `Bearer ${token}`;
  }

  return defaultOptions;
}

export default async function customFetch(url, options) {
  const response = await fetch(
    `${config.url.BACKEND_URL}${url}`,
    merge(getDefaultOptions(), options)
  );
  if (!response.ok) {
    throw new ResponseError("Bad request error", response);
  }

  return response;
}
