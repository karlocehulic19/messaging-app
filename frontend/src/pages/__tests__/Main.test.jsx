import { MemoryRouter } from "react-router-dom";
import App from "../../App";
import { act, render, screen, waitFor } from "@testing-library/react";
import "../../mocks/URL";
import { expect, vi } from "vitest";
import { userEvent } from "@testing-library/user-event";
import * as customFetch from "../../utils/customFetch";
import {
  defaultTestUser,
  firstTestUser,
  poolingTestUser,
  secondTestUser,
  TestPoolingMessage,
  oldMessagesUser,
  oldMessage,
  userWithoutPicture,
  dateMessagesUser,
  firstDateMessage,
  yesterdaysMessagesUser,
  todaysMessagesUser,
  firstYesterdaysDateMessage,
  firstTodaysDateMessage,
} from "../../mocks/handlers";
import { server } from "../../mocks/node";
import { http, HttpResponse } from "msw";
import { config, POOLING_INTERVAL_TIME_SECONDS } from "../../Constants";
import { Test2InstantMessage } from "../../mocks/handlers";

// Overrides react testing libraries tendency to use jest, needed for userEvents with fake timers
this.jest = vi;
const mockedSystemTime = "2025-04-04T20:33:37.997Z";

vi.mock("../../utils/apiErrorLogger", async (importOriginal) => ({
  default: vi.fn(async (error) => {
    if (
      ["No new messages found", "No profile picture found"].includes(
        error.message
      )
    )
      return;
    (await importOriginal()).default(error);
  }),
}));

const setup = (initialEntries = ["/"], initialIndex = 0) => {
  const user = userEvent.setup();
  localStorage.setItem("site", "randomJWTtoken");
  render(
    <App
      routerRender={(children) => (
        <MemoryRouter
          initialEntries={initialEntries}
          initialIndex={initialIndex}
        >
          {children}
        </MemoryRouter>
      )}
    />
  );

  return { user };
};

const setupMessage = async (
  initialEntries = ["/" + firstTestUser.username]
) => {
  const utils = setup(initialEntries);
  const user = utils.user;

  const messageInput = await screen.findByRole("textbox");
  const firstMessageText = "Hello World";

  await user.click(messageInput);
  await user.keyboard("Hello World");
  vi.resetAllMocks();

  async function sendMessage() {
    await user.click(screen.getByRole("button", { name: "Send button" }));
  }

  return { ...utils, user, messageInput, sendMessage, firstMessageText };
};

describe("<Main />", () => {
  it("displays selection text when no username is specified in url", () => {
    setup();
    expect(
      screen.getByRole("heading", {
        name: "Select user to message in the search bar",
      })
    ).toBeInTheDocument();
  });

  it("displays messaging interface when correct username is displayed", async () => {
    setup(["/" + firstTestUser.username]);

    expect(
      await screen.findByRole("heading", { name: "Test" })
    ).toBeInTheDocument();
    expect(
      screen.queryByRole("heading", {
        name: "Select user to message in the search bar",
      })
    ).not.toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: "Message input" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Send button" })
    ).toBeInTheDocument();

    const receiverProfilePic = await screen.findByRole("img", {
      name: `${firstTestUser.username}'s profile picture`,
    });
    expect(receiverProfilePic.src).toMatchSnapshot();
  });

  it("displays default profile picture on users without one", async () => {
    setup(["/" + secondTestUser.username]);
    vi.spyOn(console, "error").mockImplementationOnce(() => undefined);
    const receiverProfilePic = await screen.findByRole("img", {
      name: `${secondTestUser.username}'s profile picture`,
    });

    expect(receiverProfilePic.src).toMatchSnapshot();
  });

  it("displays selection text when receiver isn't real user", async () => {
    setup(["/notInDb"]);

    expect(
      screen.getByRole("heading", {
        name: "Select user to message in the search bar",
      })
    ).toBeInTheDocument();
  });

  it("sends a message using send button", async () => {
    vi.setSystemTime(mockedSystemTime);
    const { firstMessageText, messageInput, sendMessage } =
      await setupMessage();

    const customFetchSpy = vi.spyOn(customFetch, "default");

    await sendMessage();

    expect(customFetchSpy.mock.calls[0][0]).toBe("/messages");
    expect(customFetchSpy.mock.calls[0][1]).toEqual({
      method: "POST",
      body: JSON.stringify({
        sender: defaultTestUser.username,
        receiver: firstTestUser.username,
        message: firstMessageText,
        clientTimestamp: new Date(),
      }),
    });
    const message = screen.getByText(firstMessageText);
    const timeStamp = screen.getByText("22:33");
    expect(message).toBeInTheDocument();
    expect(timeStamp).toBeInTheDocument();
    expect(messageInput).toHaveValue("");
  });

  it("sends a message using enter button", async () => {
    vi.setSystemTime(mockedSystemTime);

    const { user, firstMessageText } = await setupMessage();

    const customFetchSpy = vi.spyOn(customFetch, "default");

    await user.keyboard("{Enter}");

    expect(customFetchSpy.mock.calls[0][0]).toBe("/messages");
    expect(customFetchSpy.mock.calls[0][1]).toEqual({
      method: "POST",
      body: JSON.stringify({
        sender: defaultTestUser.username,
        receiver: firstTestUser.username,
        message: firstMessageText,
        clientTimestamp: new Date(),
      }),
    });
  });

  it("doesn't send empty messages", async () => {
    setup(["/" + firstTestUser.username]);
    const user = userEvent.setup();

    await user.click(
      await screen.findByRole("button", { name: "Send button" })
    );

    expect(screen.queryByText("22:33")).not.toBeInTheDocument();
  });

  it("displays message time correctly", async () => {
    // checking for for leading zeroes in 03
    vi.setSystemTime("2025-04-04T20:03:37.997Z");
    const { user } = await setupMessage();

    await user.keyboard("{Enter}");

    expect(screen.getByText("22:03")).toBeInTheDocument();
  });

  it("displays new messages fetched when sent", async () => {
    vi.spyOn(console, "error").mockImplementationOnce(() => undefined);
    const { sendMessage, firstMessageText } = await setupMessage([
      "/" + secondTestUser.username,
    ]);

    await sendMessage();

    const newMessage = await screen.findByText(Test2InstantMessage);
    const oldMessage = screen.getByText(firstMessageText);

    expect(newMessage).toBeInTheDocument();

    const messages = screen.getAllByLabelText(/message/);
    expect(messages[0]).toBe(newMessage);
    expect(messages[1]).toBe(oldMessage);
  });

  it("displays error popup on invalid message send", async () => {
    server.use(
      http.post(`${config.url.BACKEND_URL}/messages`, () =>
        HttpResponse.error()
      )
    );
    const { sendMessage, firstMessageText } = await setupMessage();

    vi.spyOn(console, "error").mockImplementationOnce(() => undefined);
    await sendMessage();

    const errorPopup = screen.getByRole("alert");
    expect(errorPopup).toBeInTheDocument();

    expect(screen.getByRole("textbox")).toHaveValue(firstMessageText);
  });

  it("distinguishes messages from client and partner", async () => {
    vi.spyOn(console, "error").mockImplementationOnce(() => undefined);
    const { sendMessage } = await setupMessage(["/" + secondTestUser.username]);

    await sendMessage();

    await screen.findByLabelText("Partner's message");
    expect(screen.queryByLabelText("Your message")).toBeInTheDocument();
  });

  it(`pools for new messages every ${POOLING_INTERVAL_TIME_SECONDS} seconds`, async () => {
    const customFetchSpy = vi.spyOn(customFetch, "default");
    const pooledMessagesUrl = `/messages?receiver=${defaultTestUser.username}&sender=${poolingTestUser.username}`;

    vi.useRealTimers();
    vi.useFakeTimers();
    setup(["/" + poolingTestUser.username]);

    await act(() => undefined);
    expect(customFetchSpy).toHaveBeenCalledWith(pooledMessagesUrl);

    expect(screen.getByText(TestPoolingMessage)).toBeInTheDocument();

    customFetchSpy.mockClear();
    await act(() => vi.advanceTimersByTime(3000));
    expect(customFetchSpy).toHaveBeenCalledWith(pooledMessagesUrl);

    expect(screen.getAllByText(TestPoolingMessage)).toHaveLength(2);

    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it("doesn't pool if not in messaging gui", async () => {
    vi.useFakeTimers();
    setup(["/" + poolingTestUser.username]);
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
    const customFetchSpy = vi.spyOn(customFetch, "default");

    await act(() => vi.runOnlyPendingTimers());
    customFetchSpy.mockClear();
    await user.click(screen.getByLabelText("Options button"));

    await user.click(screen.getByRole("link", { name: "Settings" }));
    await act(() => vi.advanceTimersByTime(3000));

    expect(customFetchSpy).not.toHaveBeenCalledWith(
      `/messages?receiver=${defaultTestUser.username}&sender=${poolingTestUser.username}`
    );

    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it("shows older messages", async () => {
    const customFetchSpy = vi.spyOn(customFetch, "default");
    const oldMessagesUrl = `/messages/old?user=${defaultTestUser.username}&partner=${oldMessagesUser.username}`;
    setup(["/" + oldMessagesUser.username]);

    await waitFor(() => {
      expect(customFetchSpy).toHaveBeenCalledWith(
        oldMessagesUrl,
        expect.objectContaining({
          signal: expect.any(AbortSignal),
        })
      );
    });
    expect(
      customFetchSpy.mock.calls.filter(([url]) => url == oldMessagesUrl)
    ).toHaveLength(1);

    expect(screen.getByLabelText("Partner's message")).toBeInTheDocument();
  });

  it("removes messages when switching partners", async () => {
    const { user } = setup(["/" + oldMessagesUser.username]);

    await screen.findByText(oldMessage);
    await user.click(screen.getByRole("searchbox"));
    await user.keyboard(secondTestUser.username);
    await user.click(
      await screen.findByLabelText(`${secondTestUser.username} user`)
    );

    expect(screen.queryByText(oldMessage)).not.toBeInTheDocument();
  });

  it("shows default profile picture after switching from partner with profile picture", async () => {
    const { user } = setup(["/" + firstTestUser.username]);
    const defaultPictureSrc =
      "http://localhost:3000/src/assets/default-profile-picture.jpeg";

    await user.click(screen.getByRole("searchbox"));
    await user.keyboard(userWithoutPicture.username);
    await user.click(
      await screen.findByLabelText(`${userWithoutPicture.username} user`)
    );

    expect(screen.getByRole("img").src).toBe(defaultPictureSrc);
  });

  it("displays dates for older messages", async () => {
    vi.setSystemTime(mockedSystemTime);
    setup(["/" + dateMessagesUser.username]);

    await screen.findByText(firstDateMessage);
    expect(screen.getByRole("main").children).toMatchSnapshot();
  });

  it("displays yesterday tags for yesterdays messages", async () => {
    vi.setSystemTime(mockedSystemTime);
    setup(["/" + yesterdaysMessagesUser.username]);

    await screen.findByText(firstYesterdaysDateMessage);
    expect(screen.getByRole("main").children).toMatchSnapshot();
  });

  it("displays today tags for todays messages", async () => {
    vi.setSystemTime(mockedSystemTime);
    setup(["/" + todaysMessagesUser.username]);

    await screen.findByText(firstTodaysDateMessage);
    expect(screen.getByRole("main").children).toMatchSnapshot();
  });

  it("back button goes to last url not dashboard", async () => {
    vi.spyOn(window.history, "length", "get").mockReturnValueOnce(2);
    const { user } = setup(["/" + firstTestUser.username, "/settings"], 1);

    await user.click(
      await screen.findByRole("button", { name: "Back button" })
    );

    expect(
      await screen.findByRole("heading", { name: firstTestUser.username })
    ).toBeInTheDocument();
  });
});
