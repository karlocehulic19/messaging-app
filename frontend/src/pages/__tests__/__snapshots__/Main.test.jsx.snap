// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Main /> > displays dates for older messages 1`] = `
HTMLCollection [
  <span>
    Mar 30th
  </span>,
  <div
    aria-label="Partner's message"
  >
    First dated message
    <span>
      22:33
    </span>
  </div>,
  <div
    aria-label="Partner's message"
  >
    Second message sent on Sun Mar 30 2025 22:33:37 GMT+0200 (Central European Summer Time)
    <span>
      22:33
    </span>
  </div>,
  <span>
    Mar 31st
  </span>,
  <div
    aria-label="Partner's message"
  >
    Message sent on Mon Mar 31 2025 22:33:37 GMT+0200 (Central European Summer Time)
    <span>
      22:33
    </span>
  </div>,
  <span>
    Apr 1st
  </span>,
  <div
    aria-label="Partner's message"
  >
    Message sent on Tue Apr 01 2025 22:33:37 GMT+0200 (Central European Summer Time)
    <span>
      22:33
    </span>
  </div>,
]
`;

exports[`<Main /> > displays default profile picture on users without one 1`] = `"http://localhost:3000/src/assets/default-profile-picture.jpeg"`;

exports[`<Main /> > displays messaging interface when correct username is displayed 1`] = `"http://localhost:3000/mockedFilePath"`;
